---
Author: "Yogesh Budhathoki"
title: "Harmony Data Integration Tutorial"
output: 
    html_document:
        code_folding: hide
always_allow_html: true
---

# Load the library
```{r, include=FALSE}
library(rrrSingleCellUtils)
library(Seurat)
library(ggrepel)
library(tidyverse)
library(stringr)
library(harmony)
```

# Single vs combined/merged seurat objects
Single seurat objects can be plotted using the DimPlot function in Seurat. The
DimPlot function allows researchers to visualize the structure of the data in a
reduced dimensional space, such as PCA or UMAP. This can help identify cell
clusters, explore relationships between cells, and visualize the distribution of
gene expression across cells.
Merging single cell RNA-seq datasets is a common task in the analysis of
single cell data. It allows researchers to combine data from different
experiments, conditions, or technologies to increase the sample size and
improve the statistical power of the analysis.
In this tutorial, we will demonstrate how to merge two single cell RNA-seq
datasets using the MergeSeurat function in Seurat. The datasets used in this
tutorial are loaded from four different single cell experiment or samples.
The datasets are merged based on the common features and cells present in both
datasets. The merged dataset is then used for downstream analysis, such as
dimensionality reduction, clustering, and differential expression analysis.
In summary, merging single cell RNA-seq datasets is a crucial step in the
analysis of single cell data. It allows researchers to combine data from
different sources and increase the sample size for more robust and comparative
analysis under same statistical conditions.
Let's say we have four different datasets from four different samples or
experiments. We can load the datasets individually and then merge them using
the MergeSeurat function in Seurat. The merged dataset can then be used for
downstream analysis, such as dimensionality reduction, clustering, and
differential expression analysis.
If we do not merge them, we cannot perform comparative analysis between the
datasets, as they are treated as separate datasets.
```{r}
# Load the single data
control <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/S0246.qs")
timepoint1 <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/S0122.qs")
timepoint2 <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/S0067.qs")
timepoint4 <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/S0068.qs")

#merge the datasets
merged_sobj <- merge(x = control,
                     y = list(timepoint1, timepoint2, timepoint4),
                     add.cell.id = names(list(control, timepoint1, timepoint2, timepoint4)))

#another way to merge is to load them as a list and merge them
sample_names <- c("S0246", "S0122", "S0067", "S0068")

#create an empty list to store the seurat objects and later merge them
sobj_list <- list()
for (sample in sample_names) {
	object <- qs::qread(str_c("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/", sample, ".qs"))
	sobj_list[[sample]] <- object
}

sobj_list

#merge the datasets
merged_sobj <- merge(x = sobj_list[[1]],
                            y = sobj_list[2:length(sobj_list)],
                            add.cell.id = names(sobj_list))

#check if it is merged
sobj_list
merged_sobj

# reprocess the merged object together and join layers
merged_sobj <- merged_sobj %>%
	NormalizeData() %>%
	FindVariableFeatures() %>%
	ScaleData() %>%
	RunPCA() %>%
	FindNeighbors() %>%
	FindClusters() %>%
	RunUMAP(dims = 1:10) %>%
	JoinLayers()

```

# Data Visualization
Data visualization is an essential step in the analysis of single cell RNA-seq
datasets. It allows researchers to explore the structure of the data, identify
cell clusters, and visualize relationships between cells. In this tutorial, we
will demonstrate how to visualize single cell RNA-seq data using the DimPlot
function in Seurat.
The data used in this tutorial is loaded from a file named "data_dontneed_harmony.qs".
This dataset does not require Harmony integration, as it does not have significant
batch effects or other sources of variation that would necessitate data integration.
The data is visualized using the DimPlot function, with cells grouped by sample name
and Seurat clusters. This allows for the exploration of the data structure and the
identification of cell clusters without the need for data integration.
In summary, data visualization is a crucial step in the analysis of single cell RNA-seq
datasets. It provides insights into the structure of the data and helps researchers
identify cell clusters and relationships between cells.

## Data visualization using DimPlot
The DimPlot function in Seurat is used to visualize single cell RNA-seq data in a
dimensionality reduction plot. This plot displays cells in a reduced dimensional
space, such as PCA or UMAP, and allows for the exploration of cell clusters and
relationships between cells.
In the provided example, the data is loaded from a file named "data_dontneed_harmony.qs".
This dataset does not require Harmony integration, as it does not have significant
batch effects or other sources of variation that would necessitate data integration.
The data is then visualized using the DimPlot function, with cells grouped by sample
name and Seurat clusters. This allows for the exploration of the data structure and the
identification of cell clusters without the need for data integration.
In summary, the DimPlot function is a powerful tool for visualizing single cell RNA-seq
data and exploring the structure of the data. It is an essential step in the analysis
of single cell RNA-seq datasets.

```{r}
# Load the data
#save individual samples as seurat objects so that you can later merge
object <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/data_dontneed_harmony.qs")

# visualize the data
DimPlot(object,
        group.by = "seurat_clusters",
        label = TRUE,
        repel = TRUE)

# visualize the data using multiple groupings
DimPlot(object,
        group.by = c("sample_name", "seurat_clusters"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# visualize the data using specific colors
DimPlot(object,
        group.by = c("sample_name", "seurat_clusters"),
        label = TRUE,
        repel = TRUE,
        ncol = 2,
        cols = c(plot_cols, sample(rainbow(1000))))

# visualize without the legend
DimPlot(object,
        group.by = "sample_name",
        label = TRUE,
        repel = TRUE,
        cols = c(plot_cols, sample(rainbow(1000)))) +
    NoLegend() +
    coord_fixed()

# Give a title to the plot
DimPlot(object,
        group.by = "seurat_clusters",
        label = TRUE,
        repel = TRUE,
        cols = c(plot_cols, sample(rainbow(1000)))) +
    NoLegend() +
    ggtitle("Data Visualization")
```

### Quick Practice for DimPlot
practice the visualization of the data and figure out how to:
(1) change the size of each point/dot in the DimPlot
(2) change colors of the clusters
(3) change legend positions

```{r}



```


## Data Visualization using FeaturePlot
FeaturePlot is another useful function in Seurat for visualizing single cell RNA-seq
data. It allows researchers to visualize the expression of specific genes in the
dataset and explore the distribution of gene expression across cells.
In the provided example, we will use the FeaturePlot function to visualize the
expression of specific genes in the single cell RNA-seq dataset loaded from the
file "data_dontneed_harmony.qs". This dataset does not require Harmony integration,
as it does not have significant batch effects or other sources of variation that
would necessitate data integration.
The expression of specific genes is visualized using the FeaturePlot function, with
cells grouped by sample name and Seurat clusters. This allows for the exploration of
gene expression patterns across cells and the identification of genes that are
differentially expressed in specific cell clusters.
In summary, FeaturePlot is a valuable tool for visualizing gene expression patterns
in single cell RNA-seq datasets. It provides insights into the distribution of gene
expression across cells and helps researchers identify genes that are associated with
specific cell clusters.

```{r}
# Load the data
object <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/data_dontneed_harmony.qs")

# visualize the expression of specific genes
FeaturePlot(object,
            features = c("Cd3e", "Cd4", "Cd8a"))

# visualize the ncount_RNA
FeaturePlot(object,
            features = c("nCount_RNA"))

#we have our own function customized for the feature plot
r_feature_plot(object,
               features = "nCount_RNA")

# visualize the expression of specific genes using multiple groupings
FeaturePlot(object,
            features = c("Col1a1", "Cd4", "Cd8a"),
            ncol = 2)

```

### Quick Practice for FeaturePlot
practice the visualization of the data and figure out how to:
(1) Visualize gene of your interest
(2) Try identifying the potential specific celltypes (CD8, CD4) by feature plot

```{r}



```


## Data Visualization using VlnPlot
VlnPlot is a useful function in Seurat for visualizing the distribution of gene
expression across cells in a single cell RNA-seq dataset. It allows researchers to
explore the expression of specific genes and identify genes that are differentially
expressed in specific cell clusters.
In the provided example, we will use the VlnPlot function to visualize the distribution
of gene expression across cells in the single cell RNA-seq dataset loaded from the file
"data_dontneed_harmony.qs". This dataset does not require Harmony integration, as it
does not have significant batch effects or other sources of variation that would
necessitate data integration.
The distribution of gene expression is visualized using the VlnPlot function, with cells
grouped by sample name and Seurat clusters. This allows for the exploration of gene
expression patterns across cells and the identification of genes that are differentially
expressed in specific cell clusters.
In summary, VlnPlot is a valuable tool for visualizing the distribution of gene expression
across cells in single cell RNA-seq datasets. It provides insights into the expression
patterns of specific genes and helps researchers identify genes that are associated with
specific cell clusters.

```{r}

# violinplots by the ncounts, nfeatures and percent.mt 
VlnPlot(object,
        features = c("nCount_RNA", "nFeature_RNA", "percent.mt"))

# violinplots by the genes of choice
VlnPlot(control,
        features = c("Col1a1", "Col1a2", "Cd8a"),
        group.by = "seurat_clusters")
```

## Data Visualization using DotPlot
DotPlot is a useful function in Seurat for visualizing the expression of specific genes
across cell clusters in a single cell RNA-seq dataset. It allows researchers to explore
the expression patterns of genes and identify genes that are differentially expressed in
specific cell clusters.
In the provided example, we will use the DotPlot function to visualize the expression of
specific genes across cell clusters in the single cell RNA-seq dataset loaded from the
file "data_dontneed_harmony.qs". This dataset does not require Harmony integration, as it
does not have significant batch effects or other sources of variation that would
necessitate data integration.
The expression of specific genes is visualized using the DotPlot function, with cells
grouped by sample name and Seurat clusters. This allows for the exploration of gene
expression patterns across cell clusters and the identification of genes that are
differentially expressed in specific cell clusters.
In summary, DotPlot is a valuable tool for visualizing the expression of specific genes
across cell clusters in single cell RNA-seq datasets. It provides insights into the
expression patterns of genes and helps researchers identify genes that are associated
with specific cell clusters.

```{r}
# Dotplots
DotPlot(object,
        features = c("Col1a1", "Col1a2", "Cd8a"),
        group.by = "seurat_clusters",
        scale = TRUE)

# be careful with scale, it can be misleading sometimes when the scale is not the same for all the genes in the plot 
# We want scale = FALSE so that scaling more representative of the gene expression level overall
DotPlot(object,
        features = c("Col1a1", "Col1a2", "Cd8a"),
        group.by = "seurat_clusters",
        scale = FALSE)

```


## Data Visualization using RidgePlot
RidgePlot is a useful function in Seurat for visualizing the distribution of gene
expression across cell clusters in a single cell RNA-seq dataset. It allows researchers
to explore the expression patterns of genes and identify genes that are differentially
expressed in specific cell clusters.
In the provided example, we will use the RidgePlot function to visualize the distribution
of gene expression across cell clusters in the single cell RNA-seq dataset loaded from the
file "data_dontneed_harmony.qs". This dataset does not require Harmony integration, as it
does not have significant batch effects or other sources of variation that would
necessitate data integration.
The distribution of gene expression is visualized using the RidgePlot function, with cells
grouped by sample name and Seurat clusters. This allows for the exploration of gene
expression patterns across cell clusters and the identification of genes that are
differentially expressed in specific cell clusters.
In summary, RidgePlot is a valuable tool for visualizing the distribution of gene
expression across cell clusters in single cell RNA-seq datasets. It provides insights
into the expression patterns of genes and helps researchers identify genes that are
associated with specific cell clusters.

```{r}
# plots by feature of interest shows a distribution of the gene expression level
RidgePlot(object,
          features = c("Col1a1", "Col1a2", "Cd8a"),
          group.by = "sample_name")

#by seurat clusters
RidgePlot(object,
          features = c("Col1a1", "Col1a2", "Cd8a"),
          group.by = "seurat_clusters",
          sort = TRUE)

```

## Data Visualization using Heatmap
Heatmap is a useful function in Seurat for visualizing the expression of specific genes
across cell clusters in a single cell RNA-seq dataset. It allows researchers to explore
the expression patterns of genes and identify genes that are differentially expressed in
specific cell clusters.
In the provided example, we will use the Heatmap function to visualize the expression of
specific genes across cell clusters in the single cell RNA-seq dataset loaded from the
file "data_dontneed_harmony.qs". This dataset does not require Harmony integration, as it
does not have significant batch effects or other sources of variation that would
necessitate data integration.
The expression of specific genes is visualized using the Heatmap function, with cells
grouped by sample name and Seurat clusters. This allows for the exploration of gene
expression patterns across cell clusters and the identification of genes that are
differentially expressed in specific cell clusters.
In summary, Heatmap is a valuable tool for visualizing the expression of specific genes
across cell clusters in single cell RNA-seq datasets. It provides insights into the
expression patterns of genes and helps researchers identify genes that are associated
with specific cell clusters.

```{r}
# Heatmaps
DoHeatmap(object,
        features = c("Col1a1", "Col1a2", "Cd8a"))

# do expression heatmap for specific features and group by seurat clusters
DoHeatmap(object,
        features = c("Col1a1", "Col1a2", "Cd8a"),
        group.by = "sample_name",
        combine = TRUE)

# do expression heatmap for variable features
DoHeatmap(
  control,
  features =  VariableFeatures(control)[1:50],
  group.by = "seurat_clusters",
  assay = "RNA",
  slot = "scale.data",
  combine = TRUE)

DoHeatmap(object,
        features =  VariableFeatures(control)[1:50],
        group.by = "sample_name",
        assay = "RNA",
        slot = "scale.data",
        combine = TRUE)

```



# Add and Edit metadata column
Adding and editing metadata columns in a single cell RNA-seq dataset is an
important step in the analysis of single cell data. Metadata columns contain
information about the samples, cells, and experimental conditions, which can
be used for downstream analysis, visualization, and interpretation of the data.
In this tutorial, we will demonstrate how to add and edit metadata columns in
a single cell RNA-seq dataset using the AddMetaData and EditMetaData functions
in Seurat. The dataset used in this tutorial is loaded from a file named
"data.qs".
The metadata columns are added and edited based on the information provided in
the dataset. This information can include sample names, cell types, experimental
conditions, and other relevant information.
In summary, adding and editing metadata columns in a single cell RNA-seq dataset
is an essential step in the analysis of single cell data. It allows researchers
to annotate and organize the data, which can facilitate downstream analysis and
interpretation of the results.

```{r}
# add a metadata column to the seurat object
object$new_metada_column <- "value"
unique(object$new_metada_column)

# edit a metadata column in the individual seurat object by using their ID
# lets say you forgot to add the metadata column in the individual seurat object
object$new_metada_column[object$sample_name == "S0246"] <- "control"
object$new_metada_column[object$sample_name == "S0122"] <- "timpoint1"
object$new_metada_column[object$sample_name == "S0067"] <- "timepoint2"
object$new_metada_column[object$sample_name == "S0068"] <- "timpoint4"


# edit the metadata by addmetadata funtion that allows to add metadata columns by cell ids
metadata_table <- data.frame(cell_id = colnames(object),
                             new_info_column = "i_want_this_info")

# a new information has been added to the metadata table
metadata_table$new_info <- "another_info"

# add the metadata to the seurat object
object <- AddMetaData(object, metadata_table)

```


## Check cell cycle via plot
Cell cycle is an important biological process that regulates the growth and
division of cells. In single cell RNA-seq datasets, the cell cycle phase of
each cell can be inferred from the expression of cell cycle-related genes.
In this tutorial, we will demonstrate how to check the cell cycle phase of
cells in a single cell RNA-seq dataset using the CellCycleScoring function in
Seurat. The dataset used in this tutorial is loaded from a file named "data.qs".
The cell cycle phase of each cell is inferred based on the expression of cell
cycle-related genes, such as G2M and S phase genes. The CellCycleScoring
function assigns a cell cycle score to each cell, which indicates the
likelihood of the cell being in a specific cell cycle phase.
In summary, checking the cell cycle phase of cells in a single cell RNA-seq
dataset is an important step in the analysis of single cell data. It provides
insights into the biological processes that regulate cell growth and division,
and can help researchers identify cell cycle-related genes and pathways.

```{r}
# check the cell cycle phase of cells in the seurat object
# Cellcycle plots by function we developed
rrrSingleCellUtils::kill_cc(object,
                            cc = "N")

```

## RenameIdents in the seurat object
Renaming cluster names in a single cell RNA-seq dataset is an important step in
the analysis of single cell data. Cluster names provide information about the
cell types or states present in the dataset, and can help researchers interpret
the results of downstream analysis.
In this tutorial, we will demonstrate how to rename cluster names in a single
cell RNA-seq dataset using the RenameIdents function in Seurat. The dataset used
in this tutorial is loaded from a file named "data.qs".
The cluster names are changed based on the identities of the cell types or states
present in the dataset. This information can be obtained from the literature,
annotation databases, or other sources of information.
In summary, renaming cluster names in a single cell RNA-seq dataset is an
important step in the analysis of single cell data. It allows researchers to
annotate and interpret the results of downstream analysis, and provides insights
into the cell types or states present in the dataset.

```{r}
# Change cluster names - add identities
Idents(object) <- "seurat_clusters"

object_reident <- RenameIdents(object,
                                "0" = "T cells",
                                "1" = "Monocytes",
                                "2" = "NK cells",
                                "3" = "B cells",
                                "4" = "Dendritic cells",
                                "5" = "Macrophages",
                                "6" = "Fibroblasts")

DimPlot(object_reident)

```

## Subsetting the seurat object
Subsetting a single cell RNA-seq dataset is a common task in the analysis of
single cell data. It allows researchers to focus on specific cell types, states,
or conditions of interest, and perform downstream analysis on a subset of the
data.
In this tutorial, we will demonstrate how to subset a single cell RNA-seq dataset
using the subset function in Seurat. The dataset used in this tutorial is loaded
from a file named "data.qs".
The dataset is subset based on the identities of the cell types or states present
in the dataset. This information can be obtained from the cluster names, metadata
columns, or other sources of information.
In summary, subsetting a single cell RNA-seq dataset is a valuable tool for
focusing on specific cell types, states, or conditions of interest. It allows
researchers to perform targeted analysis on a subset of the data and gain insights
into the biological processes that regulate cell growth and division.

```{r}
# Subset the seurat object based on the cluster names
object_subset <- subset(object_reident,
                        idents = c("T cells", "Monocytes", "NK cells"))

# visualize the subsetted data
DimPlot(object_subset)

```

# Data integration using Harmony
## Example that do not need harmony Integration
Harmony is a powerful tool for integrating single cell RNA-seq datasets that
originate from different conditions, batches, or technologies. However, not all
datasets require this level of integration.
For instance, if a single cell RNA-seq dataset is generated under the same
conditions and using the same technology, there may be no need for Harmony
integration. This is because the batch effects that Harmony aims to correct are
not present.
In the provided example, the data is loaded from a file named
"data_dontneed_harmony.qs". This suggests that the dataset in question does not
have significant batch effects or other sources of variation that would
necessitate the use of Harmony.
The data is then visualized using a dimensionality reduction plot (DimPlot),
with cells grouped by sample name and Seurat clusters. This allows for the
exploration of the data structure and the identification of cell clusters
without the need for data integration.
In summary, while Harmony is a valuable tool for integrating diverse single
cell RNA-seq datasets, it is not always necessary. Understanding the source and
characteristics of your data is crucial in determining the appropriate analysis
steps.

```{r}
# Load the data
object <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/data_dontneed_harmony.qs")

# visualize the data combined with the sample name and seurat clusters
DimPlot(object,
        group.by = c("sample_name", "seurat_clusters"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# visualize the data combined with the sample name, seurat clusters, data source and data type
DimPlot(object,
        group.by = c("sample_name", "seurat_clusters", "data_source", "data_type"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)


```


## Simple example: data that need harmony integration
```{r}
# Load the data
simple_object <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/before_harmony_simple_example.qs")

# visualize the data
DimPlot(simple_object,
        group.by = c("sample_name", "seurat_clusters"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# visualize the data
DimPlot(simple_object,
        group.by = c("sample_name", "seurat_clusters", "model", "data_source"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# Run Harmony integration
harmony_simple_object <- 
    RunHarmony(simple_object,
               group.by.vars = "sample_name")

# input the reduction as harmony embeddings
harmony_simple_object <-
    process_seurat(harmony_simple_object,
                   reduction = "harmony")

# visualize the data
DimPlot(harmony_simple_object,
        group.by = c("sample_name", "seurat_clusters"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# visualize the data
DimPlot(harmony_simple_object,
        group.by = c("sample_name", "seurat_clusters", "model", "data_source"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

```


## Complex example of harmony integration
```{r}
# Load the data
object <- qs::qread("NCH_Coder_Upgrade/Sessions/beginner_wednesday/seurat_objects/before_harmony_complex_example.qs")

# visualize the data
DimPlot(object,
        group.by = c("sample_name", "seurat_clusters", "model", "data_source"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# Run Harmony integration
harmony_singlevars <- 
    RunHarmony(object,
               group.by.vars = "sample_name")

# input the reduction as harmony embeddings
harmony_singlevars <-
    process_seurat(harmony_singlevars,
                   reduction = "harmony")

# visualize the data
DimPlot(harmony_singlevars,
        group.by = c("sample_name", "seurat_clusters", "model", "data_source"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# did not work as good as we would like to due to a lot of batch effects
# we can try to integrate the data using the multiple factor of batch effect

# Run Harmony integration
harmony_multivars <- 
    RunHarmony(object,
               group.by.vars = c("sample_name", "data_source", "model"))

# input the reduction as harmony embeddings
harmony_multivars <-
    process_seurat(harmony_multivars,
                   reduction = "harmony")

# visualize the data
DimPlot(harmony_multivars,
        group.by = c("sample_name", "seurat_clusters", "model", "data_source"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

# didnt integrate as good as well, we can try to integrate using other featurtes of harmony

# Run Harmony integration using other features example theta
harmony_theta_features <- 
    RunHarmony(object,
               group.by.vars = c("sample_name", "data_source", "model"),
               theta = c(12, 12, 12))

# input the reduction as harmony embeddings
harmony_theta_features <-
    process_seurat(harmony_theta_features,
                   reduction = "harmony")

# visualize the data
DimPlot(harmony_theta_features,
        group.by = c("sample_name", "seurat_clusters", "model", "data_source"),
        label = TRUE,
        repel = TRUE,
        ncol = 2)

```

