---
title: "Comparing Integrative Methods"
author: "Xin Wang"
date: "2024-04-27"
output: html_document
description: This tutorial contains four methods to integrate single cell datasets. Using the IntegrateLayers function, Seurat v5 enable the streamlined integrative analysis. We hope that by simplifying the process of performing integrative analysis, you will be able to  carefully evaluate the biological information retained in the integrated dataset. For example, you can determine the best integrative method by comparing the expression of biological markers based on different clustering solutions, or visualizing one method’s clustering solution on different UMAP visualizations.
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
warnings()
```

# Loading the libraries
```{r init}
# if you have not install the software, please reinstall them into the seurat5
# remotes::install_github("satijalab/seurat", "seurat5", quiet = TRUE)
# remotes::install_github("satijalab/seurat-data", "seurat5", quiet = TRUE)
# remotes::install_github("satijalab/azimuth", "seurat5", quiet = TRUE)
# remotes::install_github("satijalab/seurat-wrappers", "seurat5", quiet = TRUE)
# remotes::install_github("stuart-lab/signac", "seurat5", quiet = TRUE)
# if the BPCells are not installed in M1 Mac, see the following https://github.com/bnprks/BPCells/issues/6
# remotes::install_github("bnprks/BPCells", quiet = TRUE)
# if you have not install the seurat data, please download and install
# devtools::install_github('satijalab/seurat-data')


# note: please confirm your R version: suggest to be R 4.3.1, given that ‘SeuratObject’ was built under R 4.3.1
# please also confirm the Matrix version: Matrix_1.6-3 

library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(ggplot2)
library(patchwork)
library(dplyr)
library(harmony)
options(future.globals.maxSize = 1e9)
sessionInfo()

```

## Introduction
Integration of single-cell sequencing datasets, for example across experimental batches, donors, or conditions, is often an important step in scRNA-seq workflows. Integrative analysis can help to match shared cell types and states across datasets, which can boost statistical power, and most importantly, facilitate accurate comparative analysis across datasets. In previous versions of Seurat we introduced methods for integrative analysis, including our ‘anchor-based’ integration workflow. Many labs have also published powerful and pioneering methods, including [Harmony](https://github.com/immunogenomics/harmony) and [scVI](https://yoseflab.github.io/software/scvi-tools/), for integrative analysis. 

We recognize that while the goal of matching shared cell types across datasets may be important for many problems, users may also be concerned about which method to use, or that integration could result in a loss of biological resolution. In Seurat v5, we introduce more flexible and streamlined infrastructure to run different integration algorithms with a single line of code. This makes it easier to explore the results of different integration methods, and to compare these results to a workflow that excludes integration steps.

For this vignette, we use a [dataset of human PBMC profiled with seven different technologies](https://www.nature.com/articles/s41587-020-0465-8), profiled as part of a systematic comparative analysis (`pbmcsca`). The data is available as part of our [SeuratData](https://github.com/satijalab/seurat-data) package. 

# Dataset introduction:
The object contains data from nine different batches (stored in the `Method` column in the object metadata), representing seven different technologies. We will aim to integrate the different batches together. In previous versions of Seurat, we would require the data to be represented as nine different Seurat objects. When using Seurat v5 assays, we can instead keep all the data in one object, but simply split the layers.

Introduction of nine different batches:
1. SMART-seq2
2. CEL-Seq2
3. 10x_Chromium_v2_A
4. 10x_Chromium_v2_B
5. 10x_Chromium_v3
6. Drop-seq # Drop-Seq analyzes mRNA transcripts from droplets of individual cells in a highly parallel fashion.
7. Seq-Well # a portable, low-cost platform for single-cell RNA sequencing designed to be compatible with low-input, clinical biopsies.
8. inDrops
9. 10x_Chromium_v2

## Layers in the Seurat v5 object
Seurat v5 assays store data in layers. These layers can store raw, un-normalized counts (`layer='counts'`), normalized data (`layer='data'`), or z-scored/variance-stabilized data (`layer='scale.data'`). We can load in the data, remove low-quality cells, and obtain predicted cell annotations (which will be useful for assessing integration later), using our [Azimuth pipeline](https://satijalab.github.io/azimuth/articles/run_azimuth_tutorial.html).

# Loading the dataset and spliting them into different layers based on methods
```{r}
# Set up your working path as where you data stored
setwd("/Users/XXW004/Documents/Projects/LectureForSingleCellGroup/2024_Update/")

# we used these datasets to build the object that contains nine different batches.
# Read the object 
obj <- readRDS(file = "IntegrativeMethods.RDS")

obj@meta.data %>% head()
# check the object and their meta data
obj # the object has two 2 layers present count and data
# let's have a look at the layers
Layers(obj[['RNA']])

# the difference between count and data sets
obj@assays$RNA$counts %>% head()#  @counts will always hold the raw counts.
obj@assays$RNA$data %>% head() # @data will hold the normalized counts if you run normalization
# check what would be the meta data of single cells
obj@meta.data %>% head()
# Check the different single cell methods:
levels(as.factor(obj@meta.data$Method))

# Split the dataset based on the Method
obj[["RNA"]] <- split(obj[["RNA"]], f = obj$Method)

# the object have been splited into 18 layers
obj
# let's have a look at the layers
Layers(obj[['RNA']])
# After splitting, there are now 18 layers (a `counts` and `data` layer for each batch). 
```

# Without any integration method, let's run a standard scRNA-seq analysis
Note that since the data is split into layers, normalization and variable feature identification is performed for each batch independently (a consensus set of variable features is automatically identified).
```{r splitassay}

obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj)

# here we can also double check the obj data and count difference
obj@assays$RNA$`counts.Smart-seq2` %>% head() # counts still store the orginal raw data
obj@assays$RNA$`data.Smart-seq2` %>% head() # data have already changed into normalized and scaled data.

```

# visualize the results of a standard analysis without integration. 
Note that cells are grouping both by cell type and by underlying method. While a UMAP analysis is just a visualization of this, clustering this dataset would return predominantly batch-specific clusters. Especially if previous cell-type annotations were not available, this would make downstream analysis extremely challenging.  

```{r unintegratedUMAP, fig.height=5, fig.width=14}

obj <- FindNeighbors(obj, dims=1:30, reduction = 'pca')
obj <- FindClusters(obj, resolution = 2, cluster.name = "unintegrated_clusters")
obj <- RunUMAP(obj, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")

# check the meta data now...
obj@meta.data %>% head()

# visualize by batch and cell cluster
DimPlot(obj, reduction = "umap.unintegrated", group.by = c("Method", "predicted.celltype.l2"))

DimPlot(object = obj, split.by = "Method",reduction = 'umap.unintegrated',group.by = c("predicted.celltype.l2"))

```

## Perform streamlined (one-line) integrative analysis

Seurat v5 enables streamlined integrative analysis using the `IntegrateLayers` function. The method currently supports four integration methods. Each of these methods performs integration in low-dimensional space, and returns a dimensional reduction (i.e. `integrated.rpca`) that aims to co-embed shared cell types across batches:

* Anchor-based CCA integration (`method=CCAIntegration`)
* Anchor-based RPCA integration (`method=RPCAIntegration`)
* Harmony (`method=HarmonyIntegration`)
* FastMNN (`method= FastMNNIntegration`)

Note that our anchor-based RPCA integration represents a faster and more conservative (less correction) method for integration. For interested users, we discuss this method in more detail in our [previous RPCA vignette](https://satijalab.org/seurat/articles/integration_rpca)

You can find more detail on each method, and any installation prerequisites, in Seurat's documentation (for example, `?scVIIntegration`). For example, scVI integration requires `reticulate` which can be installed from CRAN (`install.packages("reticulate")`) as well as `scvi-tools` and its dependencies installed in a conda environment. Please see scVI installation instructions [here](https://docs.scvi-tools.org/en/stable/installation.html).

Each of the following lines perform a new integration using a single line of code:

# Using the cca
```{r integratelayerscca, results='hide'}
obj <- IntegrateLayers(
  object = obj, method = CCAIntegration,
  orig.reduction = "pca", new.reduction = 'integrated.cca',
  verbose = FALSE)

# visualize by batch and cell cluster using the cca
DimPlot(object = obj, split.by = "Method",reduction = 'integrated.cca')

```

# Using the rpca
```{r integratelayersrpca, results='hide'}
obj <- IntegrateLayers(
  object = obj, method = RPCAIntegration,
  orig.reduction = "pca", new.reduction = 'integrated.rpca',
  verbose = FALSE)

# visualize by batch and cell cluster using the rpca
DimPlot(object = obj, split.by = "Method",reduction = 'integrated.rpca')
```

# Using the harmony
```{r integratelayersharmony, results='hide'}
obj <- IntegrateLayers(
  object = obj, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = 'harmony',
  verbose = TRUE)
# Note: you might get ~8 Warning messages, The team is still working on fixing this so the warnings wont'c come up, but this should not affect your results.

# visualize by batch and cell cluster using the harmony
DimPlot(object = obj, split.by = "Method",reduction = 'harmony')
```

# Using the FastMNN
```{r integratelayersfastmnn, results='hide'}
obj <- IntegrateLayers(
  object = obj, method = FastMNNIntegration,
  new.reduction = 'integrated.mnn',
  verbose = FALSE)
# visualize by batch and cell cluster using the FastMNN
DimPlot(object = obj, split.by = "Method",reduction = 'integrated.mnn')
```

## Then we run the seurat 
For any of the methods, we can now visualize and cluster the datasets. We show this for CCA integration and harmony, but you can do this for any method:

```{r integratedprojections, fig.height=16, fig.width=16}
# for the cca
obj <- FindNeighbors(obj, reduction = 'integrated.cca', dims = 1:30)
obj <- FindClusters(obj,resolution = 2, cluster.name = 'cca_clusters')
obj <- RunUMAP(obj, reduction = "integrated.cca", dims = 1:30, reduction.name = 'umap.cca')
p1 <- DimPlot(
  obj, reduction = "umap.cca",
  group.by = c("Method",  "cca_clusters"),
  combine = FALSE, label.size = 2)

# for the harmony
obj <- FindNeighbors(obj, reduction = 'harmony', dims = 1:30)
obj <- FindClusters(obj,resolution = 2, cluster.name = 'harmony_clusters')
obj <- RunUMAP(obj, reduction = "harmony", dims = 1:30, reduction.name = 'umap.harmony')
p2 <- DimPlot(
  obj, reduction = "umap.harmony",
  group.by = c("Method", "harmony_clusters"),
  combine = FALSE, label.size = 2) 

# wrap the plots
wrap_plots(c(p1, p2), ncol = 2, byrow = F)
```

## Potential methods to check the single cell integration.
We hope that by simplifying the process of performing integrative analysis, users can more carefully evaluate the biological information retained in the integrated dataset. For example, users can compare the expression of biological markers based on different clustering solutions, or visualize one method's clustering solution on different UMAP visualizations.

```{r vlnplots, fig.height=5, fig.width=16, warning=FALSE}
p1 <- VlnPlot(
  obj, features = "rna_CD8A", group.by = 'unintegrated_clusters'
) + NoLegend() + ggtitle("CD8A - Unintegrated Clusters")
p2 <- VlnPlot(
  obj, "rna_CD8A", group.by = 'cca_clusters'
) + NoLegend() + ggtitle("CD8A - CCA Clusters")
p3 <- VlnPlot(
  obj, "rna_CD8A", group.by = 'harmony_clusters'
) + NoLegend() + ggtitle("CD8A - harmony Clusters")
p1 | p2 | p3
```

```{r umaps, fig.height=5, fig.width=16}
obj <- RunUMAP(obj, reduction = "integrated.rpca", dims = 1:30, reduction.name = 'umap.rpca')
p4 <- DimPlot(obj, reduction="umap.unintegrated", group.by=c("cca_clusters"))
p5 <- DimPlot(obj, reduction="umap.rpca", group.by=c("cca_clusters"))
#p6 <- DimPlot(obj, reduction="umap.scvi", group.by=c("cca_clusters"))
p4 | p5 
```


#### Training session ####
## Let's integrate the COVID dataset generated yesterday using reciprocal PCA.(RPCAIntegration method in IntegrateLayers) ##
## Input dataset: 240429_merged_object.rds ##

#  Read COVID object: 240429_merged_object.rds (this object was saved from Tuesday' Seurat Part 1)
#  Integrate with RPCA 
#  Run the Seurat Pipeline to cluster the cells and non-linear dimensional reduction

```{r}
# please write your own codes here
# 
```





















```{r}
## Load the data
# Read the object 
Covidobject <- readRDS(file = "240429_merged_object.rds")

## PROCESSING
# We already performed the normalization on Tuesday, we will ignore the following pre-processes. 
# Covidobject <- NormalizeData(Covidobject)
# Covidobject <- FindVariableFeatures(Covidobject)
# Covidobject <- ScaleData(Covidobject)
# Covidobject <- RunPCA(Covidobject)

# Visualize different varieties of the single cell datasets.
DimPlot(Covidobject, group.by  = c("platform", "sampleID", "group_per_sample" ) )
Covidobject

## Using RPCAIntegration
Covidobject <- IntegrateLayers(
  object = Covidobject, method = RPCAIntegration,
  orig.reduction = "pca", new.reduction = 'integrated.rpca',
  verbose = TRUE)

# visualize by batch and cell cluster using the RPCA
DimPlot(object = Covidobject, group.by = c("platform", "sampleID", "group_per_sample" ),reduction = 'integrated.rpca')

## Run the general Seurat pipeline to find neighbor, cluster and UMAP.
Covidobject <- FindNeighbors(Covidobject, reduction = "integrated.rpca", dims = 1:30)
Covidobject <- FindClusters(Covidobject, resolution = 0.5, cluster.name = "rpca_clusters")
Covidobject <- RunUMAP(Covidobject, reduction = "integrated.rpca", dims = 1:30, reduction.name = "umap.rpca")

## Visualize the UMAP
p1 <- DimPlot(Covidobject, reduction = "umap.rpca", combine = FALSE, label.size = 2)
p1

p1_split <- DimPlot(Covidobject, reduction = "umap.rpca", split.by = c("platform"), combine = FALSE, label.size = 2)
p1_split

p1_cluster <- DimPlot(Covidobject, reduction = "umap.rpca", group.by = c("platform", "sampleID", "group_per_sample" ), combine = FALSE, label.size = 2)
wrap_plots(p1_cluster)
#wrap_plots(c(p1, p1_split,p1_cluster ), ncol = 1)
saveRDS(Covidobject, file = "integrated.rds")

# Note:
# Please save this integrated.rds object, we need this one for tomorrow's cell type annotation and differential expression section...


```



