---
title: "Session: Differential expression analysis with scRNA-seq data"
author: "Corinne Haines"
date: "2023-05-05"
output: html_document
---

# Session Description

This session will provide an introduction to differential expression analysis of
single-cell RNA-seq data.

# Session Learning Objectives

After completion, you should be able to perform differential expression analysis
using several Seurat methods and a pseudobulk method. Additionally, you should be
able to visualize the output of your differential expression analysis. Finally, 
you should have a general understanding of cell type annotation methods.

# Outline

* Load in data
    * Visualize data as a UMAP separated by treatment
* Differential gene expression between cell types
    * Change from clusters to cell type identities
    * FindAllMarkers()
    * Save markers as a .csv file
    * Visualize markers as a heatmap
    * Visualize markers as a dotplot
    * Visualize markers as a ridgeplot
* *Activity 1 (9:30-10ish)*
    * Read in data, set assay and identity
    * Visualize cell types as a UMAP split by treatment
    * Find cell type markers and visualize as dotplot
* Differential gene expression within cell types between conditions
    * Find genes differentially expressed in DC cells between ctrl and stim
    * Explore the results
    * Save DEGs as a .csv file
    * Visualize data as a volcano plot
    * Visualize select genes as a FeaturePlot
    * Visualize select genes as a VlnPlot
* Perform GSEA analysis on a list of DEGs
        * Save GSEA results as a .csv file
        * Visualize data as a bubble plot
* *Activity 2 (10:30-11ish)*
    * Find DEGs for ctrl vs. stim activated T-cells
    * Explore results and save as a .csv file
    * Visualize top 3 most significantly different genes using FeaturePlot
    * Visualize top 3 most significantly different genes using VlnPlot
    * Perform GSEA analysis on the activated T cell degs
    * Visualize GSEA results as a bubble plot
* Pseudobulked Differential Gene Expression
    * Why pseudobulk?
    * Prepare data for pseudobulk analysis
    * Perform DESeq2 on pseudobulk data
        * Save DEGs as a .csv file
        * Visualize data as a volcano plot
* *Activity 3 (11:15-11:30ish)*
    * Create pseudobulk dataset
    * Run DESeq2
    * Pull out the T activated DEGs and plot a Volcano plot
* Brief intro to cell type annotation (description)
    * Identify conserved markers for cell type assignment 
    * Use singleR
    * Annotation with other scRNA-seq datasets

* Session challenge:
    * Perform DEG analysis and visualization on clusters 
    * Perform pseudobulk analysis on clusters
    * Compare pseudobulk results to Seurat results

# Data

PBMC data from Lupus patients unstimulated or stimulated with interferon B (ifnb) (Kang et al 2017)

* Libraries were prepared using 10X Genomics v2 chemistry
* Samples were sequenced on the Illumina NextSeq 500
* PBMC samples from eight distinct lupus patients were separated into two aliquots each to receive different treatments:
    * 1 aliquot of PBMCs was activated by 100 U/mL of recombinant IFN-Î² for 6 hours
    * 1 aliquot was left untreated
* PBMCs that received the same treatment were then pooled together for processing as one 10X Genomics sample, then computationally demultiplexed:
    * After 6 hours, the eight samples for each condition were pooled together in two final pools (stimulated cells and control cells)
    * 12,138 and 12,167 cells were identified (after removing doublets) for control and stimulated pooled samples, respectively
    * Samples were demultiplexed using the tool Demuxlet
* Demo will be performed on full dataset
* Activities will be performed on subsetted dataset

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE
)

library(Seurat)
library(SeuratData)
library(ggplot2)
library(tidyverse)
library(stringr)
library(msigdbr)
library(fgsea)
library(data.table)
library(RColorBrewer)
library(ggrepel)
library(DESeq2)
library(pheatmap)


set.seed(888)


# not sure if we will need to do this or not:anchor
options(future.globals.maxSize = 8000 * 1024^24)
```

# Load in data

Load in Seurat object and set default assay to "RNA" slot
```{r load_data}
# read in seurat objected
data <- readRDS("ifnb_myeloid.rds")

# check / set default assay
DefaultAssay(data) # <- "RNA"

# what do we have in the meta data?
head(data@meta.data)

# set identity for the dataset
Idents(data) <- "seurat_annotations"

# how many cells do we have per condition?
table(data@meta.data$seurat_annotations, data@meta.data$stim)

# how many cells do we have per patient per ?
table(data@meta.data$seurat_annotations, data@meta.data$sample_ID)
```

## Visualize data as a UMAP

Set identity for Seurat object as "seurat_annotations", which is the slot in our
metadata where they cell type annotations are stored. Then plot as a UMAP 
splitting the data by stimulation status ("CTRL" or "STIM")
```{r umap}
# check active Idents for the data
head(Idents(data))
# set identity for the dataset
Idents(data) <- "seurat_annotations"
# plot UMAP separated by treatment
plot <- DimPlot(data,
                reduction = "umap",
                split.by = "stim")
plot
```


# Differential Gene Expression

In this section we will learn how to perform differential expression analysis
using Seurat built-in functions and a pseudobulk method.

## Between cell types

We can use any slot in the metadata for performing DE testing. We will start
with cell types.

### Find markers that define cell types

First, we will confirm that our identity is set to our cell type metadata slot.
We will then perform DE testing for all cell types in our data using the
FindAllMarkers function which will identify genes that are differentially
expressed in each identity group by comparing it to all of the others. Note that
this method does not identify markers that are specific to a single group, and 
thus, closely-related groups may have overlapping markers.

```{r}
# make sure identity is set to cell types
Idents(data) %>% head # <- "seurat_annotations"
# find markers for each cell type
markers <- FindAllMarkers(data,
                min.pct = 0.25) # only markers that are in 25% of cells
# reduce data
top_markers <- as.data.frame(markers %>%
                group_by(cluster) %>% # group by cluster (cell type)
                slice_max(avg_log2FC, n = 5)) # sort by log2FC and get top 5
# explore top markers
head(top_markers)
# save markers as a dataframe
write.csv(top_markers, "top5_markers.csv")
```

### Visualize markers as a heatmap

Sometimes it's nice to see globally which markers are defining specific groups 
and a heatmap is a great way to do that. We will want to use normalized counts
for this so that each group has comparable values. This will show the normalized
expression for each gene for each cell in each identity.

```{r}
# get list of top marker genes; including each gene only once
genes <- unique(top_markers$gene)
# change to transformed assay
DefaultAssay(data) <- "SCT"
# make heatmap
heatmap <- DoHeatmap(data,
                features = genes,
                group.by = "seurat_annotations",
                size = 3)
heatmap
```

### Visualize markers as a dotplot

A dot plot can also be used to look at markers for the identities. This will 
display the average expression level (color) and the percentage of cells
expressing the gene (bubble size).

```{r}
# reduce data further
top_markers <- top_markers %>%
                    group_by(cluster) %>%
                    slice_max(avg_log2FC, n = 3)
# get list of top marker genes
genes <- unique(top_markers$gene)
# make dot plot
dotplot <- DotPlot(data,
                features = genes,
                scale = FALSE) +
                RotatedAxis()
dotplot
```

### Visualize markers as a ridgeplot

Sometimes it's useful to see the distribution of expression levels within each
identity for a few genes- we can do that with a ridge plot!

```{r}
# select a few genes
genes <- c("CCL2", "TXN", "HLA-DPB1", "FCGR3A")
# make ridgeplot
ridgeplot <- RidgePlot(data,
                features = genes,
                slot = "counts",
                log = TRUE,
                ncol = 2)
ridgeplot
```

## Activity 1 (9:30-10ish)

### Read in T-cell data, set assay to the RNA slot, and set idents to *clusters*:
```{r}
tcells <- readRDS("ifnb_lymphoid.rds")
DefaultAssay(tcells) <- "RNA"
head(tcells@meta.data)
Idents(tcells) <- "seurat_clusters"
```

### Visualize *clusters* as a UMAP split by treatment
```{r}
DimPlot(tcells, 
        split.by = "stim")
```

### Find markers that define *clusters* and visualize top 2 markers using a dotplot
```{r}
# make sure identity is set to cell types
Idents(tcells) %>% head # <- "seurat_annotations"
# find markers for each cell type
markers <- FindAllMarkers(tcells,
                min.pct = 0.25) # only markers that are in 25% of cells
# reduce data
top_markers <- as.data.frame(markers %>%
                group_by(cluster) %>% # group by cluster (cell type)
                slice_max(avg_log2FC, n = 5)) # sort by log2FC and get top 5
# explore top markers
head(top_markers)
```








## Within cell types between conditions

If we have two conditions, we likely want to compare them! To do that, we can 
use the FindMarkers function.

### Find genes differentially expressed in DC cells between ctrl and stim

Here we will find the genes that are differentially expressed between CTRL and
STIM cells in the DC cell subset. We want genes that are expressed in at least
10% of the cells in each identity and have a log2FC cut-off as 0.25 (1.2X).
```{r}
Idents(data) <- "seurat_annotations"
DefaultAssay(data) <- "RNA"
# Find DEGs for DC cells
dc_degs <- FindMarkers(data,
                    ident.1 = "STIM",
                    ident.2 = "CTRL",
                    group.by = "stim",
                    subset.ident = "DC",
                    min.pct = 0.1,
                    logfc.threshold = 0.25)
```

### Explore the results
```{r}
# look at first 20 differentially expressed genes
head(dc_degs, n = 20)

# how many significantly different genes are there?
nrow(dc_degs[dc_degs$p_val_adj <= 0.05, ])
```

The results data frame has the following columns :

* p_val : p_val (unadjusted)
* avg_log2FC : log fold-change of the average expression between the two groups. 
** Note : positive values indicate that the feature is more highly expressed in the first group.
* pct.1 : percentage of cells where the feature is detected in the first group
* pct.2 : percentage of cells where the feature is detected in the second group
* p_val_adj : adjusted p-value, based on Bonferroni correction using all features in the dataset.

### Save DEGs as a .csv file

We usually want to export the files to share with others and saving them as 
a .csv file is a great way to do that.

```{r}
write.csv(dc_degs, file = "DC_stim_vs_ctrl_degs.csv")
```

### Intro to for-loops in R

In your dataset you'll likely have a few if not a dozen or more clusters and/or
cell types. Thus, instead of writing out code to do the same thing over and over
(like find DEGs) it is useful and more reproducible to iterate over elements in
a list, dataframe, vector, matrix, etc. The for loop can be used to execute a
group of statements or functions repeatedly depending on the number of elements
in your object. 

```{r}
# Example 1
fruits <- c("apple", "banana", "cherry")

for (x in fruits) {
  print(x)
}

# Example 2
my_fav_nums <- c(4,5,6)

for (x in my_fav_nums){
  y <- x + 1
  print(y)
}

```

We could also find DEGs for several cell types at a time using a for-loop

```{r}
# Find DEGs for each cell type
for (cell in unique(data$seurat_annotations)){
    # Find the markers
    markers <- FindMarkers(data,
                    ident.1 = "STIM",
                    ident.2 = "CTRL",
                    group.by = "stim",
                    subset.ident = cell,
                    min.pct = 0.1,
                    logfc.threshold = 0.25)
    # make a variable for a filename
    filename <- paste0(cell, "_STIM_vs_CTRL_DEGs.csv")
    # save our DEGs
    write.csv(markers, file=filename)
}

```

### Visualize data as a volcano plot

A volcano plot allows us to see gene expression changes and their significance 
for a lot of genes at once.

```{r}
# log transform the padj values for visualization
dc_degs$log.padj <- -log10(dc_degs$p_val_adj)
# pull out labels for the top 20 most differentially expressed genes
labels <- dc_degs %>%
                arrange(desc(abs(avg_log2FC))) %>%
                slice_head(n = 20)
# make volcano plot
volcano <- ggplot(dc_degs, aes(x = avg_log2FC, y = log.padj)) +
                geom_point() +
                geom_label_repel(data = labels,
                    aes(label = rownames(labels)),
                    max.overlaps = 20) +
                ggtitle("Dendritic Cell Differentially Expressed Genes") +
                ylab("-log(padj)") +
                xlab("log2FC") +
                theme(plot.title = element_text(hjust = 0.5),
                    text = element_text(size = 12))
volcano
```

### Visualize select genes as a FeaturePlot

A feature plot can allow us to map gene expression onto our UMAP. We can also 
split it based on a slot in the metadata, such as treatment.

```{r}
# make feature plot of one down-regulated and two up-regulated genes
feature <- FeaturePlot(data,
                features = c("IL8", "ISG15", "CXCL10"),
                split.by = "stim",
                order = TRUE) # pulls positive cells to the front
feature
```

### Visualize select genes as a VlnPlot

A violin plot allows us to see gene expression levels within each identity.

```{r}
# make violin plot
violin <- VlnPlot(data,
                features = c("IL8", "ISG15", "CXCL10"),
                split.by = "stim",
                group.by = "seurat_annotations",
                ncol = 1,
                pt.size = 0) # removes the black dots for each cell
violin
```

## Perform GSEA analysis on the DC differentially expressed genes

Gene set enrichment analysis (GSEA) helps us to see if several genes in a
particular process or pathway are changing together. We will use the
Gene Ontology Molecular Functions gene set from MSIGDB and the fgsea package.
There are several GSEA packages and gene sets you can use. You can even create 
your own gene sets!

```{r}
# look at genesets available
msigdbr_collections()
# select dataset and then aggregate the genes into a list
# each element of the list will be a named character vector
# the name of the vector is the gene set name ("gs_name")
# the elements of the vector are genes in the gene set ("gene_symbol")
go_mf <- msigdbr(species = "Homo sapiens",
                category = "C5",
                subcategory = "GO:MF") %>% 
                split(x = .$gene_symbol,
                f = .$gs_name)
# arrange DEGs for GSEA 
dc_degs$gene <- row.names(dc_degs)
deg <- dc_degs %>%
            arrange(desc(avg_log2FC)) %>% 
            dplyr::select(gene, avg_log2FC)
# convert to named list 
ranks <- deframe(deg) 
head(ranks)
# perform GSEA
fgsea_res <- fgseaMultilevel(go_mf,
                ranks,
                minSize = 15, # min size of gene set to test
                maxSize = 500, # max size of a gene set to test
                nPerm = 1000) %>% # number of permutations to test
            arrange(padj)
# take a look at results
head(fgsea_res)
# save results as a .csv file (use fwrite to handle leadingEdge list)
fwrite(fgsea_res, "DC_stim_vs_ctrl_GSEA.csv")
```

The GSEA results dataframe has the following columns:

* pathway -- name of the pathway
* pval -- an enrichment p-value
* padj -- a BH-adjusted p-value
* log2err -- the expected error for the standard deviation of the P-value
* ES -- enrichment score, same as in Broad GSEA implementation
* NES -- ES normalized to mean enrichment of random samples of the same size
* size -- size of the pathway after removing genes not present
* leadingEdge -- vector with indexes of leading edge genes that drive the 
                    enrichment

### Visualize data as a bubble plot

It's always nice to display our data as a graph! We will incorporate information
about the pathways enriched, the adjusted p-value, the size of the pathway, and
the normalized enrichment score into a bubble plot.

```{r}
# get only the top 10 pathways
fgsea_res <- fgsea_res[1:10, ]
# make bubble plot
bubble <- ggplot(fgsea_res,
                aes(x = NES,
                    y = pathway,
                    size = size,
                    color = padj)) +
                geom_point() +
                scale_color_gradient2(high = "#f517d3",
                    mid = "#ffffff",
                    low = "#0513d8",
                    midpoint = 0.05)
bubble
```


## Activity 2 (10:30-11ish)

### Find genes differentially expressed in activated T cells between ctrl and stim
```{r}

```

### Explore the results and save as a .csv file
```{r}

```

### Visualize top 3 most significantly different genes using FeaturePlot
```{r}

```

### Visualize top 3 most significantly different genes using VlnPlot
```{r}

```

## Perform GSEA analysis on the activated T cell DEGs using the Hallmark geneset
```{r}
# this gets your geneset
hallmark <- msigdbr(species = "Homo sapiens",
                category = "H") %>% 
                split(x = .$gene_symbol,
                f = .$gs_name)

```

### Visualize GSEA results as a bubble plot
```{r}

```







# Pseudobulked Differential Gene Expression
## Why pseudobulk?
The DE analysis in Seurat treats each cell as an independent sample often 
resulting in inflated p-values as single cells isolated from the same biological
sample are not *truly* independent of one another. Thus, the methods used to
perform DE analysis in Seurat identify genes that vary across individual cells 
not necessarily across a population. To infer DE genes across a population between
conditions we need to pseudobulk our cells.

## Prepare data for pseudobulk analysis
First, generate DEG tables from pseudobulk clusters, treating all cells
from each ident from each sample as independent groups of cells/data.
Then, graph each of those DEG datasets using a volcano plot.

```{r}
# Create pseudobulk tables

# set default assay to RNA
DefaultAssay(data) #<- "RNA"

# change idents to cell types
head(Idents(data)) #<- data$seurat_annotations

# make a list for pseudobulked data for each cell type to go into
pb <- list()

# for each cell type go through pseudobulking
for (cell in unique(data$seurat_annotations)) {
    # make subset for cell type
    sub <- subset(data,
            idents = cell)
    # change idents for the subtype to sample ID
    Idents(sub) <- data$sample_ID
    # get summed expression for each gene for each sample ID across all cells
    df <- AggregateExpression(sub,
            assays = "RNA",
            slot = "counts") %>%
            as.data.frame()
    # change the column names to reflect the cell type and the sample ID
    colnames(df) <- gsub("RNA.",
                        paste0(cell, "_"),
                        colnames(df))
    # move the rownames (gene names) to a column 
    df$gene <- rownames(df)
    # store the pseudobulked data into a list
    pb[[cell]] <- df
}

# merge all the pseudobulked data for each cell type into one dataframe
pb <- Reduce(merge, pb)

# make the rownames the gene names
rownames(pb) <- pb$gene

# remove the old gene name column
pb <- pb[, -c(1)]

# round the data to be integers for DESeq2
pb <- round(pb)

# can save this pseudobulked data for later
write.csv(pb, "pb.csv")

# what's the spread of gene expression for one group?
summary(pb$CD14_Mono_CTRL_1016)
hist(pb$CD14_Mono_CTRL_1016)


# Create a separate column (meta) data table as required for DESeq2 input

# make a dataframe with a sample column that has
# the columns from the pseudobulk dataframe
coldata <- data.frame(sample = colnames(pb))
coldata$sample <- gsub(" ", "_", coldata$sample) # sub spaces for "_"
# set rownames as the sample names
rownames(coldata) <- coldata$sample

# create a column that is the stim_cell type
# this one is a little wonky because of the sample names (CTRL/STIM + 3-4#s)
coldata$cell_stim <- ifelse(
    str_detect(coldata$sample, "_\\d\\d\\d\\d"), # if the sample has 4 digits
    str_remove(coldata$sample, "_\\d\\d\\d\\d"), # then remove 4 digits
    str_remove(coldata$sample, "_\\d\\d\\d")) # otherwise remove 3 digits

# make cell type column
coldata$cell_type <- ifelse(
    str_ends(coldata$cell_stim, "STIM"), # if the sample starts with STIM
    str_remove(coldata$cell_stim, "_STIM"), # remove _STIM
    str_remove(coldata$cell_stim, "_CTRL")) # otherwise remove _CTRL



# save the coldata as a .csv file for later
write.csv(coldata, "pb_coldata.csv")
```

## Perform DESeq2 between CTRL vs STIM within each cell type
```{r}
# set up DDS object using ~stim_cell as the design
dds <- DESeq2::DESeqDataSetFromMatrix(
            countData = pb,
            colData = coldata,
            design = ~ cell_stim)
# run DESeq2
dds <- DESeq2::DESeq(dds)
# save DDS object
saveRDS(dds, "dds.rds")

# make a list for the DEGs to go into
deseq2 <- list()
# pull out DEGs for each cell type
for (cell in unique(coldata$cell_type)){
    # make a variable for the ctrl cell type
    ctrl <- paste0(cell, "_CTRL")
    # make a variable for the stim cell type
    stim <- paste0(cell, "_STIM")
    # pull out the differentially expressed genes such that the stim cell type
    # is the numerator and the ctrl cell type in the denominator
    res <- DESeq2::results(dds,
                           contrast = c("cell_stim", stim, ctrl)) %>%
            as.data.frame()
    # remove any NAs in the data
    res <- res[!is.na(res$padj), ]
    # re-order the data based on padj
    res <- res[order(res$padj), ]
    # store the results in the DEG list
    deseq2[[cell]] <- res
}
# take a look at one
cell <- "pDC"
head(deseq2[[cell]])
```


### Visualize data as volcano plots
```{r}
pdc_deg <- deseq2[["pDC"]]
# pull out the names of the genes that are most significantly DE
labs <- head(pdc_deg, n = 15)
# make a violin plot for the DEGs for each cell type
plot <- ggplot(pdc_deg,
            aes(x = log2FoldChange, y = -log10(padj))) +
            geom_point() +
            geom_label_repel(
                data = labs,
                aes(label = rownames(labs)),
                max.overlaps = 20) +
            ggtitle(paste(cell, "DEGs")) +
            ylab("- log ( padj )") +
            xlab("log2FC (STIM / CTRL)") +
            theme(plot.title = element_text(hjust = 0.5),
                    text = element_text(size = 12))
plot
```

## Activity 3 (11:15-11:30ish)

### Create pseudobulk dataset for *cell types*
 
Note: if we are short on time just load in lymphoid pseudobulk data provided.

```{r}

```

### Run DESeq2
```{r}

```

### Pull out the T activated DEGs and plot a Volcano plot
```{r}

```







# Brief intro to cell type annotation (description)

## Identify conserved markers for cell type assignment

Seurat's FindConservedMarkers() function will identify genes which are
differentially expressed in each identity group (in this case, clusters) in
comparison to all other identity groups AND are similarly expressed (conserved)
between the groups you are comparing (in this case, ctrl and stim). 

We can use Seurat's FindAllMarkers() function to identify genes which are 
differentially expressed in each identity group (in this case, clusters) in 
comparison to all other identity groups. This method may identify genes which are
markers for multiple closely-related groups (i.e. not specific to one group).

You can then manually define the cluster cell type identities.

```{r manual, eval = FALSE}
# set identity of object to be clusters
Idents(data) <- "seurat_clusters"

# loop through each cluster number with FindConservativeMarkers & save data
for (i in 0:(length(levels(data$seurat_clusters)) - 1)){
    # find conserved markers for each cluster irrespective of stimulation
    conserved <- FindConservedMarkers(data, ident.1 = i, grouping.var = "stim")
    # create file name
    filename <- paste0("./conservedMarkers/Cluster", i, "_conservedMarkers.csv")
    # save markers as a .csv file
    write.csv(conserved, file = filename)
}

# run FindAllMarkers & save data for each cluster as an indiviudal file
markers <- FindAllMarkers(data)
for (i in 0:(length(levels(data$seurat_clusters)) - 1)){
    df <- markers[markers$cluster == i, ]
    filename <- paste0("./markers/Cluster", i, "_markers.csv")
    write.csv(df, file = filename)
}

# add cell type ids to metadata
Idents(data) <- "seurat_clusters"
data <- RenameIdents(data, `0` = "CD14 Mono", `1` = "CD4 Naive T",
                    `2` = "CD4 Memory T", `3` = "CD16 Mono", `4` = "B",
                    `5` = "CD8 T", `6` = "NK", `7` = "T activated", `8` = "DC",
                    `9` = "B Activated", `10` = "Mk", `11` = "pDC",
                    `12` = "Eryth", `13` = "Mono/Mk Doublets", `14` = "HSPC")
# remember to save your data!
```

## Use singleR

We can use the SingleR package and cell type references to annotate our data.

https://bioconductor.org/packages/release/bioc/html/SingleR.html

```{r singleR, eval = FALSE}
library(SingleR)
# singleR allows you to use celldex cell type references (can use others)
library(celldex)
# get count data for your scRNA-seq data
counts <- GetAssayData(data)
# load the celldex data you want to use
immgendata <- ImmGenData()
# run singleR with your count data and the cellDex data
pred <- SingleR(test = counts, ref = immgendata,
    labels = immgendata$label.fine, assay.type.test = 1)
# add the predicted cell types to your metadata
data[["ImmGen.labels"]] <- pred$labels
```

## Annotation with other scRNA-seq datasets

We can download other people's scRNA-seq data from public repositories and
use their metadata to annotate our own. This method makes use of gene set scoring.
This method can be used for cell-typing data or you could also use it for any 
other metadata column (sex, treatment, age, etc.)

```{r TransferAnchors, eval = FALSE}
# load reference from whatever source
ref <- load(coolReference.RData)
# note: reference and query dataset need to be analyzed with the same
# normalization method, so you may have to extract raw data and then
# re-normalize set identity to your goal idents
Idents(ref) <- ref$cellIDs
# visualize
DimPlot(ref, reduction = "umap")
# align assays
DefaultAssay(ref) <- "RNA"
DefaultAssay(mydata) <- "RNA"
# find anchors (use same normalization method and dims for both datasets)
anchors <- FindTransferAnchors(reference = ref,
                query = mydata,
                normalization.method = "SCT",
                dims = 1:30)
# transfer data
predictions <- TransferData(anchorset = anchors,
                    refdata = ref$cellIDs,
                    dims = 1:30)
# add predictions to metadata
mydata <- AddMetaData(mydata, metadata = predictions)
# move data to informative column
mydata$ref_labels <- mydata$predicted.id
# change idents to predicted id
Idents(mydata) <- "ref_labels"
```


# Session challenge:
## Perform DEG analysis and visualization on clusters
```{r}
```

## Perform pseudobulk and DEG analysis on clusters
```{r}
```

## How do the results of the pseudobulk analysis compare to the Seurat analysis?
```{r}
```


# Additional resources
Cell marker identification vignettes
https://satijalab.org/seurat/articles/de_vignette.html
https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md

Pseudobulk DE analysis vignette
https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/pseudobulk_DESeq2_scrnaseq.md

Single Cell Expression Atlas
https://www.ebi.ac.uk/gxa/sc/home

SingleR Github
https://github.com/dviraran/SingleR

Tutorial: guidelines for annotating single-cell transcriptomic maps using
automated and manual methods
https://www.nature.com/articles/s41596-021-00534-0

Multimodal Reference Mapping
https://satijalab.org/seurat/articles/multimodal_reference_mapping.html